pandera/dtypes_.py

```diff
@@ -0,0 +1,286 @@
+    dataclass_kwargs = {"frozen": True, "init": False, "repr": False}
+    dataclass_kwargs.update(kwargs)
+
+    if dtype is None:
+        return functools.partial(dataclass, **dataclass_kwargs)
+    return dataclass(**dataclass_kwargs)(dtype)
+
+
+class DisableInitMixin:
+    def __init__(self) -> None:
+        pass
+
+
+class DataType:
+    def __init__(self):
+        if self.__class__ is DataType:
+            raise TypeError(
+                f"{self.__class__.__name__} may not be instantiated."
+            )
+
+    def __call__(self, obj: Any):
+        """Coerce object to the dtype."""
+        return self.coerce(obj)
+
+    def coerce(self, obj: Any):
+        """Coerce object to the dtype."""
+        raise NotImplementedError()
+
+    def __repr__(self) -> str:
+        return f"DataType({str(self)})"
+
+    def __str__(self) -> str:
+        """Must be implemented by subclasses."""
+        raise NotImplementedError()
+
+    def check(self, datatype: "DataType") -> bool:
+        if not isinstance(datatype, DataType):
+            return False
+        return self == datatype
+
+    def __hash__(self) -> int:
+        pass
+
+
+################################################################################
+# boolean
+################################################################################
+
+
+@immutable
+class Bool(DataType):
+    """Semantic representation of a boolean data type."""
+
+    def __str__(self) -> str:
+        return "bool"
+
+
+Boolean = Bool
+
+################################################################################
+# number
+################################################################################
+
+
+@immutable
+class _Number(DataType):
+    continuous: bool = None
+    exact: bool = None
+
+    def check(self, datatype: "DataType") -> bool:
+        if self.__class__ is _Number:
+            return isinstance(datatype, (Int, Float, Complex))
+        return super().check(datatype)
+
+
+@immutable
+class _PhysicalNumber(_Number):
+    bit_width: int = None
+    _base_name: str = field(default=None, init=False, repr=False)
+
+    def __eq__(self, obj: object) -> bool:
+        if isinstance(obj, type(self)):
+            return obj.bit_width == self.bit_width
+        return super().__eq__(obj)
+
+    def __str__(self) -> str:
+        return f"{self._base_name}{self.bit_width}"
+
+
+################################################################################
+## signed integer
+################################################################################
+
+
+@immutable(eq=False)
+class Int(_PhysicalNumber):
+    _base_name = "int"
+    continuous = False
+    exact = True
+    bit_width = 64
+    signed: bool = field(default=True, init=False)
+
+
+@immutable
+class Int64(Int, _PhysicalNumber):
+    bit_width = 64
+
+
+@immutable
+class Int32(Int64):
+    bit_width = 32
+
+
+@immutable
+class Int16(Int32):
+    bit_width = 16
+
+
+@immutable
+class Int8(Int16):
```

Comment:
Can you clarify the purpose of this inheritance chain? e.g. can we have all the `Int*` types inherit from `Int`?

Reply:
We could use the chain to soften the dtype check, i.e. allowing a subtype (inspired by [numpy.can_cast](https://numpy.org/doc/stable/reference/generated/numpy.can_cast.html#numpy.can_cast). I played with a `casting` argument in `DataType.check` that could be False for strict check (current behavior) or True to allow safe downcasting. Then I realized almost anything can be casted to a string and gave up the idea for now.